File: src/solutions/python/SerializeandDeserializeBinaryTree.py
from typing import List, Optional
import collections

'''
# Problem
 	- `Link`: https://leetcode.com/problems/serialize-and-deserialize-binary-tree/
# Solution
	- `Author`: Kyungtaek Lim (Jonas)
	- `Date`: July 3, 2025
	- `Answer`: (serializeBfs, deserializeBfs) / (serializeBfsDifferent, deserializeBfsDifferent) / (serializeDfsRecursive, deserializeDfsRecursive) / (serializeDfsBetter, deserializeDfsBetter)
'''

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:
    
    '''
    # Option #1
    - BFS
    - O(n)
    '''
    def serializeBfs(self, root: TreeNode) -> str:
        
        result: List[str] = []

        deque = collections.deque()
        deque.append(root)

        while deque:
            node: TreeNode = deque.popleft()
            if not node:
                result.append("N")
                continue

            result.append(str(node.val))
            deque.append(node.left)
            deque.append(node.right)

        return ",".join(result)
    

    def deserializeBfs(self, data: str) -> TreeNode:
        
        list: List[str] = data.split(",")
        if list[0] == "N":
            return None
        
        deque = collections.deque()
        root = TreeNode(int(list[0]))
        deque.append(root)
        i: int = 1

        while deque:
            node: TreeNode = deque.popleft()
            
            if list[i] != "N":
                node.left = TreeNode(int(list[i]))
                deque.append(node.left)
            i += 1

            if list[i] != "N":
                node.right = TreeNode(int(list[i]))
                deque.append(node.right)
            i += 1
        
        return root

    
    '''
    # Option #2
    - BFS Different
    - O(n)
    - August 25, 2026
    '''
    def serializeBfsDifferent(self, root: Optional[TreeNode]) -> str:
        if not root:
            return "null"
        
        queue = collections.deque([root])
        result: List[str] = []
        while queue:
            n = len(queue)
            for i in range(n):
                node = queue.popleft()
                if not node:
                    result.append("null")
                else:
                    result.append(str(node.val))
                    queue.append(node.left)
                    queue.append(node.right)
        return ",".join(result)


    def deserializeBfsDifferent(self, data: str) -> Optional[TreeNode]:
        if data == "null":
            return None

        strings: List[str] = data.split(",")
        root: Optional[TreeNode] = TreeNode(int(strings[0]))
        queue = [root]
        head = 0
        
        for i in range(1, len(strings)):
            node: Optional[TreeNode] = None
            if strings[i] != "null":
                node = TreeNode(int(strings[i]))
                queue.append(node)
            
            if i % 2 == 1:
                queue[head].left = node
            else:
                queue[head].right = node
                head += 1
        
        return root
            
    
    '''
    # Option #3
    - DFS Recursive
    - O(n)
    '''
    def serializeDfsRecursive(self, root: TreeNode) -> str:
        result: List[str] = []
        
        def dfs(node: TreeNode):
            if not node:
                result.append("N")
                return
            
            result.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
        
        dfs(root)
        return ",".join(result)

    def deserializeDfsRecursive(self, data: str) -> TreeNode:
        list: List[str] = data.split(",")
        self.i = 0
        
        def dfs() -> TreeNode:
            if list[self.i] == "N":
                self.i += 1
                return None
            
            node = TreeNode(int(list[self.i]))
            self.i += 1

            node.left = dfs()
            node.right = dfs()
            return node

        return dfs()
            

    
    '''
    # Option #4
    - DFS Recursive Better
    - O(n)
    - August 25, 2026
    '''
    def serializeDfsBetter(self, root: Optional[TreeNode]) -> str:
        if not root:
            return "null"
        left, right = self.serializeDfsBetter(root.left), self.serializeDfsBetter(root.right)
        return f"{root.val},{left},{right}"
        

    def deserializeDfsBetter(self, data: str) -> Optional[TreeNode]:
        if data == "null":
            return None
        strings = data.split(",")
        i = [0]

        def dfs() -> Optional[TreeNode]:
            if i[0] >= len(strings):
                return None

            if strings[i[0]] == "null":
                i[0] += 1
                return None

            node = TreeNode(int(strings[i[0]]))
            i[0] += 1
            node.left = dfs()
            node.right = dfs()
            return node

        return dfs()
        
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))



if __name__ == "__main__":

    # [1,2,3,null,null,4,5]
    root: TreeNode = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.right.left = TreeNode(4)
    root.right.right = TreeNode(5)
    print(root.val)
    print(root.left.val)
    print(root.right.val)
    print(root.right.left.val)
    print(root.right.right.val)

    sol = Codec()
    string = sol.serialize(root=root)

    print(string)

    root = sol.deserialize(data=string)
    print(root.val)
    print(root.left.val)
    print(root.right.val)
    print(root.right.left.val)
    print(root.right.right.val)

File: src/solutions/java/SerializeandDeserializeBinaryTree.java
package solutions.java;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/*
# Problem
 	- `Link`: https://leetcode.com/problems/serialize-and-deserialize-binary-tree/
# Solution
	- `Author`: Kyungtaek Lim (Jonas)
	- `Date`: July 8, 2025
	- `Answer`: (serializeBfs, deserializeBfs) / (serializeBfsBetter, deserializeBfsBetter) / (serializeDfsRecursive, deserializeDfsRecursive) / (serializeDfsBetter, deserializeDfsBetter)
 */
public class SerializeandDeserializeBinaryTree {
	
	public static class TreeNode {
	    int val;
	    TreeNode left;
	    TreeNode right;
	    TreeNode(int x) { val = x; }
	}
	
	public static class Codec {
		
		
		/*
		 * =================================================================
	    # Option #1
	    - BFS
	    - O(n)
		 */
		final static int INVALID_VAL = 1001;
		
	    // Encodes a tree to a single string.
	    public String serializeBfs(TreeNode root) {
	    	if (root == null) return "N";
	    	
	        StringBuilder result = new StringBuilder();

	        Deque<TreeNode> deque = new ArrayDeque<>();
	        deque.offer(root);
	        
	        while (!deque.isEmpty()) {
	        	TreeNode node = deque.pollFirst();
	        	if (node.val == INVALID_VAL) {
	        		result.append("N,");
	        	} else {
	        		result.append(node.val).append(",");
	        		if (node.left != null) deque.offer(node.left);
	        		else deque.offer(new TreeNode(INVALID_VAL));
	        		
	        		if (node.right != null) deque.offer(node.right);
	        		else deque.offer(new TreeNode(INVALID_VAL));
	        	}
	        }
	        
	        return result.toString();
	    }

	    // Decodes your encoded data to tree.
	    public TreeNode deserializeBfs(String data) {
	        
	    	String[] nodes = data.split(",");
	    	if ("N".equals(nodes[0])) return null;
	    	
	    	Deque<TreeNode> deque = new ArrayDeque<>();
	    	TreeNode root = new TreeNode(Integer.parseInt(nodes[0]));
	    	deque.offer(root);
	    	int i = 1;
	    	
	    	while (!deque.isEmpty()) {
	    		TreeNode node = deque.pollFirst();
	    		
	    		String left = nodes[i++];
	    		if (!left.isEmpty() && !"N".equals(left)) {
	    			node.left = new TreeNode(Integer.parseInt(left));
	    			deque.offer(node.left);
	    		}
	    		
	    		String right = nodes[i++];
	    		if (!left.isEmpty() &&!"N".equals(right)) {
	    			node.right = new TreeNode(Integer.parseInt(right));
	    			deque.offer(node.right);
	    		}
	    	}
	    	
	    	return root;
	    }

		
		/*
		 * =================================================================
	    # Option #2
	    - Bfs Better
	    - O(n)
		- August 25, 2026
		 */
		// Encodes a tree to a single string.
		public String serializeBfsBetter(TreeNode root) {
			if (root == null) return "null";
			
			Deque<Object> deque = new ArrayDeque<>();
			deque.offer(root);
			List<String> result = new ArrayList<>();

			while (!deque.isEmpty()) {
				int n = deque.size();
				for (int i = 0; i < n; i++) {
					Object obj = deque.pollFirst();
					if (obj instanceof Integer) {
						result.add("null");
					} else {
						TreeNode node = (TreeNode)obj;
						result.add("" + node.val);
						deque.offer(node.left == null ? 0 : node.left);
						deque.offer(node.right == null ? 0 : node.right);
					}
				}
			}

			return String.join(",", result);
		}

		// Decodes your encoded data to tree.
		public TreeNode deserializeBfsBetter(String data) {
			if (data.equals("null")) return null;
			
			String[] strings = data.split(",");
			List<TreeNode> list = new ArrayList<>();
			TreeNode root = new TreeNode(Integer.valueOf(strings[0]));
			list.add(root);
			int head = 0;

			for (int i = 1; i < strings.length; i++) {

				TreeNode node = null;
				if (!strings[i].equals("null")) {
					node = new TreeNode(Integer.valueOf(strings[i]));
					list.add(node);
				}
				
				if (i % 2 == 1) {
					list.get(head).left = node;
				} else {
					list.get(head++).right = node;
				}
			}

			return root;
		}


		
		/*
		 * =================================================================
	    # Option #3
	    - DFS
	    - O(n)
		 */
	    // Encodes a tree to a single string.
	    public String serializeDfsRecursive(TreeNode root) {
	    	StringBuilder result = new StringBuilder();
	    	dfsForSerializing(root, result);
	    	return result.toString();
	    }
	    
	    private void dfsForSerializing(TreeNode node, StringBuilder result) {
	    	if (node == null) {
	    		result.append("N,");
	    		return;
	    	}
	    	result.append(node.val).append(",");
	    	dfsForSerializing(node.left, result);
	    	dfsForSerializing(node.right, result);
	    }
	    
	    static int INDEX = 0;

	    // Decodes your encoded data to tree.
	    public TreeNode deserializeDfsRecursive(String data) {
	    	INDEX = 0;
	    	return dfsForDeserializing(data.split(","));
	    }
	    
	    private TreeNode dfsForDeserializing(String[] nodes) {
	    	if (INDEX >= nodes.length) return null;
	    	String nodeValue = nodes[INDEX++];
	    	if (nodeValue.isEmpty() || nodeValue.equals("N")) return null;
	    	
	    	TreeNode node = new TreeNode(Integer.parseInt(nodeValue));
	    	node.left = dfsForDeserializing(nodes);
	    	node.right = dfsForDeserializing(nodes);
	    	return node;
	    }


		
		/*
		 * =================================================================
	    # Option #4
	    - DFS Better
	    - O(n)
		 */
		// Encodes a tree to a single string.
		public String serializeDfsBetter(TreeNode root) {
			if (root == null) return "null";
			String left = serializeDfsBetter(root.left);
			String right = serializeDfsBetter(root.right);
			return root.val + "," + left + "," + right;
		}

		// Decodes your encoded data to tree.
		public TreeNode deserializeDfsBetter(String data) {
			if (data.equals("null")) return null;
			String[] strings = data.split(",");
			int[] i = new int[] {0};
			return dfs(strings, i);
		}

		private TreeNode dfs(String[] strings, int[] i) {
			if (i[0] >= strings.length) return null;
			if (strings[i[0]].equals("null")) {
				i[0]++;
				return null;
			}
			TreeNode node = new TreeNode(Integer.valueOf(strings[i[0]]));
			i[0]++;
			node.left = dfs(strings, i);
			node.right = dfs(strings, i);
			return node;
		}
	}

	public static void main(String[] args) {
		
		 Codec ser = new Codec();
		 Codec deser = new Codec();
		 
		 // [1,2,3,null,null,4,5]
		 TreeNode root = new TreeNode(1);
		 root.left = new TreeNode(2);
		 root.right = new TreeNode(3);
		 root.right.left = new TreeNode(4);
		 root.right.right = new TreeNode(5);
		 
		 String serialized = ser.serializeBfs(root);
		 System.out.println(serialized);
		 
		 TreeNode ans = deser.deserializeBfs(serialized);
		 System.out.println(ans.val);
		 System.out.println(ans.left.val);
		 System.out.println(ans.right.val);
		 System.out.println(ans.right.left.val);
		 System.out.println(ans.right.right.val);
		 

		 
		 // [1,2,3,null,null,4,5]
		 root = new TreeNode(1);
		 root.left = new TreeNode(2);
		 root.right = new TreeNode(3);
		 root.right.left = new TreeNode(4);
		 root.right.right = new TreeNode(5);
		 
		 serialized = ser.serializeDfsRecursive(root);
		 System.out.println(serialized);
		 
		 ans = deser.deserializeDfsRecursive(serialized);
		 System.out.println(ans.val);
		 System.out.println(ans.left.val);
		 System.out.println(ans.right.val);
		 System.out.println(ans.right.left.val);
		 System.out.println(ans.right.right.val);
		 
		 
		 
		 // [1,2,3,null,null,4,5]
		 root = new TreeNode(1);
		 
		 serialized = ser.serializeDfsRecursive(root);
		 System.out.println(serialized);
		 
		 ans = deser.deserializeDfsRecursive(serialized);
		 System.out.println(ans.val);
	}
}
